### ISSUE TITLE
eval 'require Foo::Bar' differs from require Foo::Bar

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6404]
Migrated from [rt.perl.org#21742](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=21742) (status was 'resolved')

Searchable as RT21742$

**Comment by p5pRT** [comment]
### From @schwern

#### Created by @schwern



$ perl5\.8\.0 \-e 'require Pod&#8203;::Man;  print $@&#8203;'
$ perl5\.8\.0 \-e 'eval q\[require Pod&#8203;::Man\];  print $@&#8203;'
Pod/Man\.pm did not return a true value at \(eval 1\) line 3\.

This problem occurs in at least 5\.8\.0 and bleadperl@&#8203;18374\.  It is not
in 5\.6\.0\.

Its pretty serious\.  eval q\[require Foo&#8203;::Bar\] is the usual way to
check if one has a module installed and causes MakeMaker to incorrectly
assume a module is not installed\.  It only appears to be a problem if
the module does not have an explicit true value at the end\.

Some discussion about the bug is in this thread&#8203;:
http&#8203;://archive\.develooper\.com/vmsperl@&#8203;perl\.org/msg06569\.html

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.0:

Configured by schwern at Sat Feb 22 11:32:44 GMT 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=darwin, osvers=6.4, archname=darwin-64all
    uname='darwin schwerns-computer.local. 6.4 darwin kernel version 6.4: wed jan 29 18:50:42 pst 2003; root:xnuxnu-344.26.obj~1release_ppc power macintosh powerpc '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-pipe -fno-common -no-cpp-precomp -DDEBUGGING -fno-strict-aliasing',
    optimize='-O3 -g',
    cppflags='-no-cpp-precomp -pipe -fno-common -no-cpp-precomp -DDEBUGGING -fno-strict-aliasing'
    ccversion='', gccversion='3.1 20020420 (prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=87654321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-flat_namespace -L/sw/lib'
    libpth=/sw/lib /usr/lib
    libs=-lgdbm -ldb -lm -lc
    perllibs=-lm -lc
    libc=/usr/lib/libc.dylib, so=dylib, useshrplib=true, libperl=libperl.dylib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dyld.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -flat_namespace -bundle -undefined suppress -L/sw/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /sw/lib/perl5
    /usr/local/perl5.8.0/lib/darwin-64all
    /usr/local/perl5.8.0/lib
    /usr/local/perl5.8.0/lib/site_perl/darwin-64all
    /usr/local/perl5.8.0/lib/site_perl
    /usr/local/perl5.8.0/lib/site_perl
    /usr/local/perl5.8.0/lib/vendor/darwin-64all
    /usr/local/perl5.8.0/lib/vendor
    /usr/local/perl5.8.0/lib/vendor
    .


Environment for perl v5.8.0:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/schwern
    LANG=en_US
    LANGUAGE=en
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/schwern/bin:/usr/local/bin:/sw/bin:/sw/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin
    PERL5LIB=/sw/lib/perl5
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From dankogai@dan.co.jp

On Saturday\, Mar 29\, 2003\, at 14&#8203;:38 Asia/Tokyo\, Michael G Schwern \(via 
RT\) wrote&#8203;:
> $ perl5\.8\.0 \-e 'require Pod&#8203;::Man;  print $@&#8203;'
> $ perl5\.8\.0 \-e 'eval q\[require Pod&#8203;::Man\];  print $@&#8203;'
> Pod/Man\.pm did not return a true value at \(eval 1\) line 3\.
>
> This problem occurs in at least 5\.8\.0 and bleadperl@&#8203;18374\.  It is not
> in 5\.6\.0\.
>
> Its pretty serious\.  eval q\[require Foo&#8203;::Bar\] is the usual way to
> check if one has a module installed and causes MakeMaker to incorrectly
> assume a module is not installed\.  It only appears to be a problem if
> the module does not have an explicit true value at the end\.
>
> Some discussion about the bug is in this thread&#8203;:
> http&#8203;://archive\.develooper\.com/vmsperl@&#8203;perl\.org/msg06569\.html

I thinks this is unique to Pod&#8203;::Man;

 > perl5\.8\.0  \-e 'eval q\[require Encode\];  print $@&#8203;'
 > perl5\.8\.0 \-e 'eval q\[require Digest&#8203;::MD5\];  print $@&#8203;'
 > perl5\.8\.0  \-e 'eval q\[require Encode\];  print $@&#8203;'
 > perl5\.8\.0 \-e 'eval q\[require Digest&#8203;::MD5\];  print $@&#8203;'
 > perl5\.8\.0 \-e 'eval q\[require Time&#8203;::HiRes\];  print $@&#8203;'

They all appear normal\.  Then I took a look at Pod&#8203;::Man and found there 
was no idiomatic "1;" before "\_\_END\_\_"\. I made a local copy of Pod&#8203;::Man 
and aplied the patch below\.  Then

perl5\.8\.0 \-I\. \-e 'eval q\[require Pod&#8203;::Man\];  print $@&#8203;; print 
$INC\{"Pod/Man\.pm"\}'

happily prints "Pod/Man\.pm"\.

&nbsp;  diff \-u  /usr/local/lib/perl5/5\.8\.0/Pod/Man\.pm Pod/Man\.pm
<details><summary>Inline Patch</summary>

```diff
--- /usr/local/lib/perl5/5.8.0/Pod/Man.pm       Wed Jan 15 23:11:52 2003
+++ Pod/Man.pm  Sat Mar 29 14:55:42 2003
@@ -1121,6 +1121,7 @@
      }
  }

+1; # So perl and Schwern is happy
  __END__


Dan the 
```
</details>

**Comment by p5pRT** [comment]
### From @schwern

On Sat\, Mar 29\, 2003 at 03&#8203;:02&#8203;:13PM \+0900\, Dan Kogai wrote&#8203;:
> On Saturday\, Mar 29\, 2003\, at 14&#8203;:38 Asia/Tokyo\, Michael G Schwern \(via 
> RT\) wrote&#8203;:
> >$ perl5\.8\.0 \-e 'require Pod&#8203;::Man;  print $@&#8203;'
> >$ perl5\.8\.0 \-e 'eval q\[require Pod&#8203;::Man\];  print $@&#8203;'
> >Pod/Man\.pm did not return a true value at \(eval 1\) line 3\.

\<snip>

> They all appear normal\.  Then I took a look at Pod&#8203;::Man and found there 
> was no idiomatic "1;" before "\_\_END\_\_"\. I made a local copy of Pod&#8203;::Man 
> and aplied the patch below\.  Then
> 
> perl5\.8\.0 \-I\. \-e 'eval q\[require Pod&#8203;::Man\];  print $@&#8203;; print 
> $INC\{"Pod/Man\.pm"\}'
> 
> happily prints "Pod/Man\.pm"\.

No\, that is merely covering up the problem\.

&nbsp;   require Pod&#8203;::Man;

and

&nbsp;   eval q\[require Pod&#8203;::Man\];

should be functionally equivalent\.  In 5\.8\.0 they are not\, the one in the
eval is very touchy about wanting the final "1;"\.  To illustrate further\,
there is this work around&#8203;:

&nbsp;   eval q\[require Pod&#8203;::Man; 0\];

that one will happily load Pod&#8203;::Man without the final "1;"\.  This is
definately an eval bug\.


\-\- 
List context isn't dangerous\.  Misquoting Gibson is dangerous\.
&nbsp;        \-\- Ziggy

**Comment by p5pRT** [comment]
### From rra@stanford.edu

Michael G Schwern \<schwern@&#8203;pobox\.com> writes&#8203;:

> No\, that is merely covering up the problem\.

>    require Pod&#8203;::Man;

> and

>    eval q\[require Pod&#8203;::Man\];

> should be functionally equivalent\.  In 5\.8\.0 they are not\, the one in the
> eval is very touchy about wanting the final "1;"\.  To illustrate further\,
> there is this work around&#8203;:

>    eval q\[require Pod&#8203;::Man; 0\];

> that one will happily load Pod&#8203;::Man without the final "1;"\.  This is
> definately an eval bug\.

Works fine with 5\.6\.1&#8203;:

windlord&#8203;:\~> perl \-e 'eval q\[require Pod&#8203;::Man\]; print $@&#8203;'
windlord&#8203;:\~> 

so I'm inclined to agree\.  This is the first time I've had any reports of
this being a problem\.  I forget why I didn't bother to include the 1; I
think someone told me a while back that it was actually cargo cult
programming and I left it out and noticed it still worked fine everywhere\.

\-\- 
Russ Allbery \(rra@&#8203;stanford\.edu\)             \<http&#8203;://www\.eyrie\.org/~eagle/>

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
@rgs - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Change \#19126 was a poor attempt at fixing this\.
Check out ticket \#22708 and change \#19801 for more info\.

**Comment by p5pRT** [comment]
### From @rgs

This patch seems to fix bug \#21742\, about forcing require\(\) to be
always called in scalar context\, even from within an eval\(""\)\.
Since I'm quite late here\, I'm posting it\, so little gnomes will
find bugs in it while I'm asleep\. The trick is the check the
eval'd op tree to see if it ends with a require\.

Index&#8203;: pp\_ctl\.c
===================================================================
\-\-\- pp\_ctl\.c    \(revision 1626\)
\+\+\+ pp\_ctl\.c    \(working copy\)
@&#8203;@&#8203; \-2781\,7 \+2781\,7 @&#8203;@&#8203;
 S\_doeval\(pTHX\_ int gimme\, OP\*\* startop\, CV\* outside\, U32 seq\)
 \{
&nbsp;     dSP;
\-    OP \*saveop = PL\_op;
\+    OP \*saveop = PL\_op\, \*seqop;
 
&nbsp;     PL\_in\_eval = \(\(saveop && saveop\->op\_type == OP\_REQUIRE\)
&nbsp;           ? \(EVAL\_INREQUIRE | \(PL\_in\_eval & EVAL\_INEVAL\)\)
@&#8203;@&#8203; \-2829\,7 \+2829\,6 @&#8203;@&#8203;
&nbsp;     sv\_setpv\(ERRSV\,""\);
&nbsp;     if \(yyparse\(\) || PL\_error\_count || \!PL\_eval\_root\) \{
&nbsp;     SV \*\*newsp;
\-    I32 gimme;
&nbsp;     PERL\_CONTEXT \*cx;
&nbsp;     I32 optype = 0;            /\* Might be reset by POPEVAL\. \*/
&nbsp;     STRLEN n\_a;
@&#8203;@&#8203; \-2873\,7 \+2872\,12 @&#8203;@&#8203;
&nbsp;     \*startop = PL\_eval\_root;
&nbsp;     \} else
&nbsp;     SAVEFREEOP\(PL\_eval\_root\);
\-    if \(gimme & G\_VOID\)
\+
\+    seqop = PL\_eval\_root ? \(\(UNOP\*\)PL\_eval\_root\)\->op\_first : Nullop;
\+    if \(seqop && seqop\->op\_type == OP\_LINESEQ
\+        && \(\(LISTOP\*\)seqop\)\->op\_last\->op\_type == OP\_REQUIRE\)
\+    scalar\(PL\_eval\_root\);
\+    else if \(gimme & G\_VOID\)
&nbsp;     scalarvoid\(PL\_eval\_root\);
&nbsp;     else if \(gimme & G\_ARRAY\)
&nbsp;     list\(PL\_eval\_root\);
Index&#8203;: t/comp/require\.t
===================================================================
\-\-\- t/comp/require\.t    \(revision 1626\)
\+\+\+ t/comp/require\.t    \(working copy\)
@&#8203;@&#8203; \-11\,7 \+11\,7 @&#8203;@&#8203;
 
 my $Is\_EBCDIC = \(ord\('A'\) == 193\) ? 1 : 0;
 my $Is\_UTF8   = \($\{^OPEN\} || ""\) =~ /&#8203;:utf8/;
\-my $total\_tests = 29;
\+my $total\_tests = 30;
 if \($Is\_EBCDIC || $Is\_UTF8\) \{ $total\_tests = 26; \}
 print "1\.\.$total\_tests\\n";
 
@&#8203;@&#8203; \-134\,8 \+134\,7 @&#8203;@&#8203;
 
 write\_file\('bleah\.pm'\, \<\<'\*\*BLEAH\*\*'
 print "not " if \!defined wantarray || wantarray ne '';
\-my $TODO = $i == 23 ? " \# TODO bug \#21742" : "";
\-print "ok $i \- require\(\) context$TODO\\n";
\+print "ok $i \- require\(\) context\\n";
 1;
 \*\*BLEAH\*\*
 \);
@&#8203;@&#8203; \-143\,6 \+142\,7 @&#8203;@&#8203;
 $foo = eval q\{require bleah\}; delete $INC\{"bleah\.pm"\}; \+\+$&#8203;::i;
 @&#8203;foo = eval q\{require bleah\}; delete $INC\{"bleah\.pm"\}; \+\+$&#8203;::i;
&nbsp;        eval q\{require bleah\}; delete $INC\{"bleah\.pm"\}; \+\+$&#8203;::i;
\+       eval q\{$\_=1\+2;require bleah\}; delete $INC\{"bleah\.pm"\}; \+\+$&#8203;::i;
 $foo = eval  \{require bleah\}; delete $INC\{"bleah\.pm"\}; \+\+$&#8203;::i;
 @&#8203;foo = eval  \{require bleah\}; delete $INC\{"bleah\.pm"\}; \+\+$&#8203;::i;
&nbsp;        eval  \{require bleah\};

**Comment by p5pRT** [comment]
### From @rgs

Fixed in bleadperl by change \#19851\, unless proven otherwise\, of
course\. I'm marking it as resolved\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
